src/uvarint.cr

```diff
@@ -3,42 +3,112 @@ require "./uvarint/*"
 # This module is based on the varint implementation in the Go programming
 # language. See: https://golang.org/src/encoding/binary/varint.go
 
-MSB = 0x80_u8  # 10000000
-REST = 0x7F_u8 # 01111111
-
-module UVarint
-  extend self
-
-  alias UVarInt = Array(UInt8)
+# Encodes an Int::Unsined into a Bytes slice.
+private def encode(n : Int::Unsigned) : Bytes
+  ptr = Pointer(UInt8).malloc 10
+  i = 0
+  while n >= 0x80_u8
+    ptr[i] = n.to_u8 | 0x80_u8
+    i += 1
+    n = n >> 7
+  end
+  ptr[i] = n.to_u8
+  Bytes.new(ptr, i + 1)
+end
 
-  # encode encodes a UInt64 into a UVarInt,
-  # which is an Array of UInt8's.
-  def encode(n : UInt64) : UVarInt
-    buf = [] of UInt8
-    while n >= MSB
-      buf << (n.to_u8 | MSB)
-      n = n >> 7
-    end
-    buf << n.to_u8
-    return buf
-  end
-
-  # decode decodes a UVarInt into a UInt64.
-  # If the UVarInt contains more than 10 UInt8's,
-  # then an overflow exception is raised.
-  def decode(vi : UVarInt) : UInt64
-    x = 0_u64
-    s = 0
-    vi.each_with_index do |v, i|
-      if v < MSB
-        if i > 9 || i == 9 && v > 1
-          raise Exception.new("overflow")
-        end
-        return x | (v.to_u64 << s)
+# Decodes enumerable bytes into a UInt64.
+# If the enumeration is longer thar 10 bytes,
+# then an overflow exception is raised.
+private def decode(bytes : Bytes) : UInt64
+  x = 0_u64
+  s = 0
+  bytes.each_with_index do |b, i|
+    if b < 0x80_u8
+      if i > 9 || i == 9 && b > 1
+        raise Exception.new "overflow"
       end
-      x |= (v & REST).to_u64 << s
-      s += 7
+      return x | (b.to_u64 << s)
     end
-    return 0_u64
+    x |= (b & 0x7F_u8).to_u64 << s
+    s += 7
+  end
+  0_u64
+end
+
+struct UVarInt
+  @bytes : Bytes
```

Comment:
Why are we storing both the `@bytes` and `@uint`? The `UInt64` is a poor choice because it has an inherent limit, which is what uvarint is supposed to avoid. Also, if both these members are supposed to represent the same information, then keeping both may likely lead to inconsistent state.